depthai_ros_driver/include/depthai_ros_driver/device_ros.hpp

```diff
@@ -150,19 +150,23 @@ class DeviceROS : public dai::DeviceBase {
 
 protected:
     // create stream based on name at the time of subscription
-    template <class MsgType, dai::DatatypeEnum DataType>
-    auto generate_cb_lambda(std::unique_ptr<dai::XLinkStream>& stream, msgpack::sbuffer& sbuf,
-            std::vector<std::uint8_t>& writer_buf) -> boost::function<void(const boost::shared_ptr<MsgType const>&)> {
-        const auto core_sub_lambda = [&stream, &sbuf, &writer_buf](const boost::shared_ptr<MsgType const>& msg) {
+    template <class MsgType>
+    auto generate_cb_lambda(const std::string& name) -> boost::function<void(const boost::shared_ptr<MsgType const>&)> {
+                std::unique_ptr<dai::XLinkStream>& stream = _streams[name];
+                std::vector<std::uint8_t>& serial_buf = _serial_bufs[name];
+                std::vector<std::uint8_t>& writer_buf = _writer_bufs[name];
+
+        const auto core_sub_lambda = [&stream, &serial_buf, &writer_buf](const boost::shared_ptr<MsgType const>& msg) {
             Guard guard([] { ROS_ERROR("Communication failed: Device error or misconfiguration."); });
 
             // convert msg to data
-            msgpack::pack(sbuf, *msg);
-            PacketWriter writer(*stream);
-            writer.write(msg->data.data(), msg->data.size(), reinterpret_cast<std::uint8_t*>(sbuf.data()), sbuf.size(),
-                    static_cast<std::uint32_t>(DataType), writer_buf);
+            dai::DatatypeEnum datatype;
+            auto data_ref = adapt_ros2dai<MsgType>::convert(*msg);
+            data_ref.serialize(serial_buf, datatype);
```

Comment:
Doesn't this call the serialize function from dai?

Reply:
the binary data might be necessary to be in the format of dai json.
serialized binary data size for `RawCameraControl` are 
```
dai serialization -> 349 bytes
msgpack packing -> 37 bytes
```
whereas `sizeof(RawCameraControl)  = 96 bytes` and the data packed are `59` bytes.

Reply:
Maybe we need to add MSGPACK_MAP or related macro. That seems to be the cause for reduction in size. I'll ask Luxonis about the serialization used for confirmation